File: mobile-app/src/Services/CommentService.ts
import { api } from './HttpClient';

export interface Comment {
    id: string;
    content: string;
    type?: "TEXT" | "GIF";
    likeCount: number;
    username: string;
    avatarUrl?: string;
    createdAt: string;
    updatedAt: string;
    isLikedByCurrentUser?: boolean;  // Add this field from backend
    timeAgo?: string;  // Thời gian đã format từ backend: "vừa xong", "3 phút trước", "25/7"
    replyCount?: number;  // Số lượng replies của comment
    parentCommentId?: string; // ID của comment cha (nếu có)
}

export interface CommentCreateResponse {
    comment: Comment;
    totalComments: number;
}

export interface CommentActionResponse {
    liked: boolean;
    disliked: boolean;
    likeCount: number;
    dislikeCount: number;
}

export interface CommentPageResponse {
    content: Comment[];
    totalElements: number;
    totalPages: number;
    size: number;
    number: number;
}

class CommentService {
    async addComment(
        feedItemId: string,
        content: string,
        parentCommentId?: string,
        type: "TEXT" | "GIF" = "TEXT"
    ): Promise<CommentCreateResponse> {
        const response = await api.post(
            `/feed-items/${feedItemId}/comments`,
            {
                content,
                type,
                ...(parentCommentId && { parentCommentId })
            }
        );
        // Backend wraps response in { code, result, message }
        return response.data.result || response.data;
    }

    /**
     * Reply to a comment
     * @param feedItemId - ID of the feed item
     * @param parentCommentId - ID of the parent comment to reply to
     * @param content - Reply content
     * @returns CommentCreateResponse
     */
    async replyComment(
        feedItemId: string,
        parentCommentId: string,
        content: string,
        type: "TEXT" | "GIF" = "TEXT"
    ): Promise<CommentCreateResponse> {
        return this.addComment(feedItemId, content, parentCommentId, type);
    }

    async getComments(
        feedItemId: string,
        page: number = 0,
        size: number = 20
    ): Promise<CommentPageResponse> {
        const response = await api.get(
            `/feed-items/${feedItemId}/comments`,
            {
                params: {
                    page,
                    size,
                },
            }
        );
        // Backend wraps response in { code, result, message }
        return response.data.result || response.data;
    }

    /**
     * Like a comment
     * @param commentId - ID của comment cần like
     * @returns CommentActionResponse với trạng thái và số lượng likes mới
     */
    async likeComment(commentId: string): Promise<CommentActionResponse> {
        const response = await api.post(
            `/feed-items/comments/${commentId}/like`
        );
        return response.data.result || response.data;
    }

    /**
     * Unlike a comment
     * @param commentId - ID của comment cần unlike
     * @returns CommentActionResponse với trạng thái và số lượng likes mới
     */
    async unlikeComment(commentId: string): Promise<CommentActionResponse> {
        const response = await api.delete(
            `/feed-items/comments/${commentId}/like`
        );
        return response.data.result || response.data;
    }

    /**
     * Toggle like status of a comment (like nếu chưa like, unlike nếu đã like)
     * @param commentId - ID của comment
     * @param isCurrentlyLiked - Trạng thái like hiện tại
     * @returns CommentActionResponse với trạng thái và số lượng likes mới
     */
    async toggleCommentLike(
        commentId: string,
        isCurrentlyLiked: boolean
    ): Promise<CommentActionResponse> {
        if (isCurrentlyLiked) {
            return this.unlikeComment(commentId);
        } else {
            return this.likeComment(commentId);
        }
    }

    async removeComment(commentId: string): Promise<boolean> {
        const response = await api.delete(
            `/feed-items/comments/${commentId}`
        );
        return response.data.result || response.data;
    }

    /**
     * Get replies of a comment
     * @param commentId - ID of the parent comment
     * @param page - Page number (default: 0)
     * @param size - Page size (default: 20)
     * @returns CommentPageResponse
     */
    async getReplies(
        commentId: string,
        page: number = 0,
        size: number = 20
    ): Promise<CommentPageResponse> {
        const response = await api.get(
            `/feed-items/comments/${commentId}/replies`,
            {
                params: {
                    page,
                    size,
                },
            }
        );
        return response.data.result || response.data;
    }
}

export default new CommentService();

File: backend/src/main/java/com/hehe/thesocial/service/metadata/comment/CommentService.java
package com.hehe.thesocial.service.metadata.comment;

import com.hehe.thesocial.dto.request.comment.CommentCreateRequest;
import com.hehe.thesocial.dto.response.comment.CommentCreateResponse;
import com.hehe.thesocial.dto.response.comment.CommentResponse;
import com.hehe.thesocial.dto.response.metadata.CommentActionResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface CommentService {
    /**
     * Thêm comment mới cho một FeedItem
     * @param feedItemId - ID của FeedItem cần thêm comment
     * @param request - Thông tin comment (content, userDetailId, etc.)
     * @return CommentCreateResponse chứa comment vừa tạo và tổng số comments
     */
    CommentCreateResponse addComment(String feedItemId, CommentCreateRequest request);

    /**
     * Xóa comment khỏi FeedItem
     * @param commentId - ID của comment cần xóa
     * @param userDetailId - ID của user đang xóa comment (để check quyền)
     * @return true nếu xóa thành công
     */
    boolean removeComment(String commentId, String userDetailId);

    /**
     * Like một comment
     * @param commentId - ID của comment cần like
     * @param userDetailId - ID của user đang like
     * @return CommentActionResponse chứa trạng thái liked và số lượng like mới
     */
    CommentActionResponse likeComment(String commentId, String userDetailId);

    /**
     * Bỏ like một comment
     * @param commentId - ID của comment cần unlike
     * @param userDetailId - ID của user đang unlike
     * @return CommentActionResponse chứa trạng thái liked và số lượng like còn lại
     */
    CommentActionResponse unlikeComment(String commentId, String userDetailId);

    /**
     * Dislike một comment
     * @param commentId - ID của comment cần dislike
     * @param userDetailId - ID của user đang dislike
     * @return CommentActionResponse chứa trạng thái disliked và số lượng dislike mới
     */
    CommentActionResponse dislikeComment(String commentId, String userDetailId);

    /**
     * Bỏ dislike một comment
     * @param commentId - ID của comment cần undislike
     * @param userDetailId - ID của user đang undislike
     * @return CommentActionResponse chứa trạng thái disliked và số lượng dislike còn lại
     */
    CommentActionResponse undislikeComment(String commentId, String userDetailId);

    /**
     * Lấy thông tin chi tiết của một comment
     * @param commentId - ID của comment
     * @param currentUserDetailId - ID của user hiện tại (để check liked/disliked status)
     * @return CommentResponse
     */
    CommentResponse getComment(String commentId, String currentUserDetailId);

    /**
     * Lấy danh sách comments của một FeedItem
     * @param feedItemId - ID của FeedItem
     * @param currentUserDetailId - ID của user hiện tại
     * @param page - Trang hiện tại (bắt đầu từ 0)
     * @param size - Số lượng comment trên mỗi trang
     * @return Page<CommentResponse>
     */
    Page<CommentResponse> getCommentsByFeedItem(
            String feedItemId,
            String currentUserDetailId,
            Pageable pageable
    );
}
